### PR TITLE
implement ui and interfaces

### PR DESCRIPTION
**Pull request by MarsLviv**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,15 @@
+*.iml
+.gradle
+/local.properties
+/.idea/caches
+/.idea/libraries
+/.idea/modules.xml
+/.idea/workspace.xml
+/.idea/navEditor.xml
+/.idea/assetWizardSettings.xml
```
Comment (gutachkee) [inline .gitignore]: I would rather ignore all Idea files. It likes to do a lot of changes there periodically. No need to store them in VCS  
```suggestion
.idea
```

File: app/build.gradle [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    compileOptions {
+        sourceCompatibility JavaVersion.VERSION_1_8
+        targetCompatibility JavaVersion.VERSION_1_8
+    }
+}
+
+dependencies {
+
```
Comment (gutachkee) [inline app/build.gradle]:
```suggestion
```

File: app/src/main/jni/def.h [hunk 3]
```diff
@@ -0,0 +1,19 @@
+#include <android/log.h>
+#include <string.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#ifndef PICTUREN_DEF_H
+#define PICTUREN_DEF_H
+
+#ifdef __ANDROID__
```
Comment (gutachkee) [inline app/src/main/jni/def.h]: if you're using checks for \_\_ANDROID\_\_, then I believe you should put `#include <android/log.h>` line inside ifdef.

From the other hand, we do not expect this code to be built without NDK, so there is no need to have it(ifdef) at all. 
(But we would need to have it is case when we want the code to be compiled without NDK, i.e. for some other platforms)

File: app/src/test/java/com/example/picturen/ExampleUnitTest.java [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+ * Example local unit test, which will execute on the development machine (host).
+ *
+ * @see <a href="http://d.android.com/tools/testing">Testing documentation</a>
+ */
+public class ExampleUnitTest {
+    @Test
+    public void addition_isCorrect() {
+        assertEquals(4, 2 + 2);
+    }
+}
```
Comment (gutachkee) [inline app/src/test/java/com/example/picturen/ExampleUnitTest.java:17]: We will modify it for our needs in future. But lets keep it as is for now.

File: app/src/main/jni/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,11 @@
+cmake_minimum_required(VERSION 3.4.1)
+
+add_library(capture
+            SHARED
+            capture.cpp )
```
Comment (gutachkee) [inline app/src/main/jni/CMakeLists.txt]: formatting
```suggestion
add_library(
        capture
        SHARED
        capture.cpp
)
```

File: app/src/main/jni/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+
+find_library(log
+             log )
+
+target_link_libraries(  capture
+                        log )
```
Comment (gutachkee) [inline app/src/main/jni/CMakeLists.txt]: again: formatting.
findLibrary with log log?
```suggestion
find_library(
        log
)

target_link_libraries(
        capture
        log
)
```
ALWAYS DO SELF REVIEW BEFORE PUSHING THE CHANGE

File: app/src/main/jni/def.h [hunk 7]
```diff
@@ -0,0 +1,19 @@
+#include <android/log.h>
+#include <string.h>
+#include <stdio.h>
+#include <stdlib.h>
+
```
Comment (gutachkee) [inline app/src/main/jni/def.h]: Let's have this file as log_utils.h.

What 'def.h' means?

File: app/src/main/java/com/example/picturen/SavePicture.java [hunk 8]
```diff
@@ -0,0 +1,5 @@
+package com.example.picturen;
+
+public interface SavePicture {
```
Comment (gutachkee) [inline app/src/main/java/com/example/picturen/SavePicture.java]: Remove entire file.

**Review by gutachkee — CHANGES_REQUESTED** [review]
There are no required changes.
The task was:
1. Create UI. Layout is defined, but no Java code is present. Button click handling? EditText value read? I expect this PR to contain all code related to the UI.
2. Define JNI interface. Interface which will be used for capturing the frame. But there is no way how to receive frame from the native. No implementation is required, just the interface.

Few things to learn:
1. Always do self review before pushing the change. This PR contain only few small  files created by yourself and they contain fomatting issues, deps duplicates..
2. If the task is not clear, **you should always ask for clarification.**

File: app/src/main/java/com/example/picturen/SavePicture.java (hunk 8 above)
Comment (MarsLviv) [reply app/src/main/java/com/example/picturen/SavePicture.java]: In order to pursue Dependency inversion principle I created this interface.
Implementation classes will save picture on Disk, on Cloud starage, on  social services, etc

File: app/src/main/jni/CMakeLists.txt (hunk 6 above)
Comment (MarsLviv) [reply app/src/main/jni/CMakeLists.txt]: find_library command require at least two parameters

File: app/src/main/java/com/example/picturen/FrameWriterSD.java [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+import android.util.Log;
+
+import androidx.annotation.NonNull;
+
+public class FrameWriterSD implements FrameWriter {
+    private static final String TAG = FrameWriterSD.class.getSimpleName();
+    @Override
+    public void write(@NonNull byte[] frame, @NonNull FrameWriterCallback saveCallback) {
+        Log.i(TAG, "save");
+        saveCallback.onSave();
```
Comment (gutachkee) [inline app/src/main/java/com/example/picturen/FrameWriterSD.java]:
```suggestion
        // todo: do the implementation
        saveCallback.onSave();
```

File: app/src/main/java/com/example/picturen/NativeCamera.java [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+    private static final String TAG = NativeCamera.class.getSimpleName();
+
+    static { System.loadLibrary("capture-lib"); }
+
+    public native byte[] capture();
+
+    @Override
+    public void capture(@NonNull CameraCallback captureCallback) {
+        Log.i(TAG, "capture");
+        byte[] frame = (byte[]) capture();
```
Comment (gutachkee) [inline app/src/main/java/com/example/picturen/NativeCamera.java]:
```suggestion
        byte[] frame = capture();
```

**Review by gutachkee — APPROVED** [review]
@MarsLviv please, resolve few additional comments and feel free to merge.

File: app/src/main/java/com/example/picturen/FrameWriterSD.java (hunk 9 above)
Comment (MarsLviv) [reply app/src/main/java/com/example/picturen/FrameWriterSD.java]: done

File: app/src/main/java/com/example/picturen/NativeCamera.java (hunk 10 above)
Comment (MarsLviv) [reply app/src/main/java/com/example/picturen/NativeCamera.java]: done

**Pull request merged**